### ISSUE TITLE
Missing LINEST and SLOPE Formula Support

### ISSUE THREAD

**Issue by omegaroy** [ClosedXML/ClosedXML#1828]
## Read and complete the full issue template

Do not randomly delete sections. They are here for a reason.

**Do you want to request a *feature* or report a *bug*?**
- [ ] Bug
- [x] Feature
- [ ] Question

**Did you test against the latest CI build?**
- [x] Yes
- [ ] No

If you answered `No`, please test with [the latest development build](https://ci.appveyor.com/project/ClosedXML/ClosedXML/branch/develop/artifacts) first.

**Version of ClosedXML**

0.95.4

**What is the current behavior?**

Missing LINEST and SLOPE functions.

**What is the expected behavior or new feature?**

Documentation says to let you know if missing functions are needed, so would like LINEST and SLOPE functions added.

**Is this a regression from the previous version?**

No

## Reproducibility
**This is an important section. Read it carefully. Failure to do so will cause a 'RTFM' comment.**

N/A

**Code to reproduce problem:**

N/A

- [ ] I attached a sample spreadsheet.  (You can drag files on to this issue)

**Comment by ghcaveman** [comment]
We have the same issue, and require similar functionality.  Rather than open a new issue, I am piggy-backing upon this one.  We would be exceptionally thankful for any efforts towards resolving this issue for us.

## Read and complete the full issue template

Do not randomly delete sections. They are here for a reason.

**Do you want to request a *feature* or report a *bug*?**
- [] Bug
- [X ] Feature
- [ ] Question

**Did you test against the latest CI build?**
- [X] Yes
- [ ] No

If you answered `No`, please test with [the latest development build](https://ci.appveyor.com/project/ClosedXML/ClosedXML/branch/develop/artifacts) first.

**Version of ClosedXML**

0.96.0

**What is the current behavior?**

An exception is thrown if a function that uses Linest is included somewhere: ClosedXML.Excel.CalcEngine.Exceptions.NameNotRecognizedException' occurred in ClosedXML.dll " +
            "The identifier `LINEST` was not recognised.

**What is the expected behavior or new feature?**

The function should provide expected behavior

**Is this a regression from the previous version?**

This problem is evident in earlier revisions - we initially saw the problem in 0.95.0 and an upgrade to 0.96.0 did not address the issue.

## Reproducibility

**Code to reproduce problem:**
```c#
public void Main()
{
      string lcFilename = "LR_SN202204-000075Edited.xlsx";

      Debug.WriteLine(lcFilename);

      XLWorkbook lcWorkbook = new XLWorkbook(lcFilename);

      IXLWorksheet lcWorksheet = lcWorkbook.Worksheet("Compliance Sheet");

      Int32 liCol = 11;
      int[] liRows = { 12, 13, 14 };

      for (int i = 0; i < liRows.Length; i++)
      {
        Int32 liCurrentRow = liRows[i];
        bool leWorked = lcWorksheet.Cell(liCurrentRow, liCol).TryGetValue<string>(out string lcCellValue);

        if (leWorked == true)
        {
          double.TryParse(lcCellValue, out double lfValue);
          string lcMessage = "Cell at Row " + liRows[i].ToString() + ", Column 11 Passed";
          Debug.WriteLine("\n" + lcMessage + "\n" + "Cell Value = " + lcCellValue.ToString());
          Assert.Greater(lfValue, 95, "Value should be 95 or greater");
        }
        else if (leWorked == false)
        {
          string lcMessage = "Cell at Row " + liRows[i].ToString() + ", Column 11 Failed";
          Debug.WriteLine("\n" + lcMessage);
          try
          {
            //this causes an exception if the ClosedXML doesn't handle LINEST
            Debug.WriteLine(lcWorksheet.Cell(liCurrentRow, liCol).Value);
          }
          catch
          {
            string lcErrorMessage = "An unhandled exception of type 'ClosedXML.Excel.CalcEngine.Exceptions.NameNotRecognizedException' occurred in ClosedXML.dll " +
            "The identifier `LINEST` was not recognised.";
            Debug.WriteLine(lcErrorMessage);
          }
        }
      }
}
```

- [ X] I attached a sample spreadsheet.  (You can drag files on to this issue)
[LR_SN202204-000075Edited.xlsx](https://github.com/ClosedXML/ClosedXML/files/9161971/LR_SN202204-000075Edited.xlsx)

**Comment by jahav** [comment]
Hi, LINEST returns an array, so ClosedXml it must be able to use array formulas. That can be done only after #1812. 

New parser should land 0.97, new CalcEngine (evaluates formulas) is half finished. Then we need to add support for array formulas/dynamic array formulas (=get result from CalcEngine and put it into correct cells in the worksheet). It's one of two priorites, so it should be soonish (month or two for PR). There is no use to work on functions before then.

**Comment by jahav** [comment]
@igitur What do you think about using libraries for functions? Although we could implement various math functions, or (better yet) offload it to someone who requests it, it would be easier (and probably better, since they have more optimized code) to use a library (e.g. Math.NET) and write a function as an adapter. Either directly in the core package or as an add-on package with a registration call from an external package. I would like to keep core package reasonably clean&lean, so I think add-on is a better option. Or we could have a simple, non-optimized implementations in the core and fast, optional add-ons. That is probably best, but I think my time is better spent elsewhere = community contribution for such implementations (after we finish How-To-Implement-Function-For-ClosedXml guide).

Similar story for financial functions (project [ExcelFinancialFunctions](https://github.com/fsprojects/ExcelFinancialFunctions), though that is pretty lean (70 KiB) library itself with no dependencies. 

Even after we complete basic CalcEngine redesign, I don't want to expose public formula value types yet (there will be a various changes and tweaks of structures). In the longer run, I want to enable custom functions, just not yet.
